### PR TITLE
Added new settings 'current_price_oracles' and 'historical_price_oracles'

### LINKED ISSUE #2131: Allow user to customize order of price querying oracles
## Abstract

At the moment both for historical prices and for current prices we check:

1. Cryptocompare
2. If that fails we check Coingecko
3. If that also fails we bail

This is a nice way but not failsafe as we just saw with NXM current price being wrong in cryptocompare but fine in coingecko. Resulting in all users getting the wrong price of NXM.

## Task

- [ ] Allow customization of order or price oracles query.
- [ ] Have a different setting (value can be the same) for current price and for historical price

### PR DESCRIPTION
**Pull request by vnavascues**
~~Historical price oracles implementation. Let's see how the API tests behave~~

Closes #2131

### CODE REVIEW

File: rotkehlchen/externalapis/coingecko.py
Comment (vnavascues) [inline rotkehlchen/externalapis/coingecko.py:257]: I've safely renamed it, it wasn't directly called in the tests.

File: rotkehlchen/inquirer.py
Comment (vnavascues) [inline rotkehlchen/inquirer.py:343]: Standardised method for querying the current USD price via current price oracles.
It is backed by a test that checks all current price oracles implement this method.

File: rotkehlchen/history/price.py
Comment (vnavascues) [inline rotkehlchen/history/price.py:207]: Standardised method for querying a historical price via historical price oracles.
It is backed by a test that checks all historical price oracles implement this method.

File: rotkehlchen/tests/unit/test_price_historian.py
Comment (vnavascues) [inline rotkehlchen/tests/unit/test_price_historian.py:108]: If you find a class within a test ugly, I can try to fake it in another way. Same for the test below.
I wasn't sure how many times I was gonna need and with which properties, so I opted for make it work first.

File: rotkehlchen/history/typing.py
Comment (vnavascues) [inline rotkehlchen/history/typing.py:14]: I think this abstraction will be useful for having more granularity over the oracle calls later on. At least I hope so.

File: rotkehlchen/inquirer.py
Comment (vnavascues) [inline rotkehlchen/inquirer.py:335]: A test checks that all the supported oracles have this method implemented.

File: rotkehlchen/history/price.py
Comment (vnavascues) [inline rotkehlchen/history/price.py:199]: A test checks that all the supported oracles have this method implemented.

File: rotkehlchen/externalapis/coingecko.py
Comment (vnavascues) [inline rotkehlchen/externalapis/coingecko.py:299]: I've added this method for being able to standardise the calls across the current and historical price oracles.

**Review by vnavascues — COMMENTED** [review]
@LefterisJP this is ready to be reviewed. I leave it as WIP as I want to check how the integration tests behave and I must fix an annotation in an overload method (I left a comment).

As always let me know your thoughts

File: rotkehlchen/rotkehlchen.py
Comment (vnavascues) [reply rotkehlchen/rotkehlchen.py:966]: Do I need to create a NewType for it and remove Type? I'm going to try it

File: rotkehlchen/inquirer.py
Comment (vnavascues) [inline rotkehlchen/inquirer.py:302]: I've just realised that preventing the extra queries on the new current price endpoint is gonna be much easier after merging this PR (again. if the proposed solution here is good).

Now that `find_usd_price` calls:
```python
price = oracle_instance.query_current_price(
    from_asset=asset,
    to_asset=A_USD,
 )
```

What I will do either in another PR or in this one:
1. Rename `find_usd_price` as `find_price` enabling `to_asset` argument. Also remove 'USD' from its logs and use `to_asset.identifier`.
2. Amend the cache for storing tuples as key: `(from_asset, to_asset)`.
3. Create a new `find_usd_price` that will call `find_price(from_asset=asset, to_asset=A_USD, ignore_cache=ignore_cache)`

File: rotkehlchen/externalapis/coingecko.py
Comment (vnavascues) [reply rotkehlchen/externalapis/coingecko.py:299]: I've seen PR #2183 adds and relies on `can_query_history`. I'll need to take it into account

File: rotkehlchen/history/price.py
Comment (vnavascues) [reply rotkehlchen/history/price.py:199]: I've seen PR #2183 adds and relies on `can_query_history`. I'll need to take it into account.

Have you considered creating an interface for price oracles? Is it worth it?

Comment (LefterisJP) [reply rotkehlchen/history/price.py:199]: not yet

File: rotkehlchen/rotkehlchen.py
Comment (LefterisJP) [reply rotkehlchen/rotkehlchen.py:966]: One mistake you have here is the return type

```diff
-    ) -> List[Union[CurrentPriceOracle, HistoricalPriceOracle]]:
+    ) -> Union[List[CurrentPriceOracle], List[HistoricalPriceOracle]]:
```
You can never have a list with a mist of different types. But even with this you get

```
Incompatible return value type (got "List[Union[CurrentPriceOracle, HistoricalPriceOracle]]", expected "Union[List[CurrentPriceOracle], List[HistoricalPriceOracle]]")
```

Which leads me to think that this may be a problem since it can't understand that `price_oracle_type` can only be one of the two types and not both. It can't determine what type it can be so it takes the union which fucks up the typing here.

File: rotkehlchen/rotkehlchen.py
Comment (LefterisJP) [inline rotkehlchen/rotkehlchen.py:994]: if you do what I told you in encoding an enum value would come here and not a string. And this should make things easier in all these functions.

File: rotkehlchen/tests/api/test_settings.py
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_settings.py:602]: As I have said before for such easy tests it's better to just unit test. Bringing up the entire api server may be a bit of an overkill.

At least parametrize the eth accounts to be zero (they are not by default) so all ethereum intiialization stuff does not happen.

File: rotkehlchen/tests/unit/test_inquirer.py
Comment (LefterisJP) [inline rotkehlchen/tests/unit/test_inquirer.py:121]: if the default order changes this test will fail by doing this. Why are you doing this? `_cryptocompare` is the object attribute the test is patching here. This is not changing.

File: rotkehlchen/tests/unit/test_price_historian.py
Comment (LefterisJP) [inline rotkehlchen/tests/unit/test_price_historian.py:28]: This is only for this file, right? You are not exporting it anywhere else? Cause we already got a price historian fixture in `tests/fixtures/history.py`

File: rotkehlchen/tests/unit/test_price_historian.py
Comment (LefterisJP) [reply rotkehlchen/tests/unit/test_price_historian.py:108]: All this to test the thing I told you to at the end remove.

 And yes this test is super ugly and lots of unneeded code. You can just mock `Inquirer().query_historical_fiat_exchange_rates` alone. Just like we do with so many other functions around the tests.

File: rotkehlchen/db/settings.py
Comment (LefterisJP) [inline rotkehlchen/db/settings.py:65]: If you follow what I said in encoding these would become a list of enum values. And you would have to serialize/deserialize before writing/reading from the DB. Do that.

**Review by LefterisJP — COMMENTED** [review]
Okay there is few nice things in here but overall it needs a ton of changes.

And it already has a lot of tests which will also need to be deleted/rewritten.

So @vnavascues I will say again what I have said many times in the past.

:clap:  **PLEASE** :clap:   (1) ask for reviews earlier (2) keep it as simple as possible (3) write no tests before I have confirmed on the approach. Why? Reasons below:

- To not waste your time. You spent quite some time on this PR working on it few days on and off. You went ahead and made some things without asking for my feedback. Then you also spent more time writing tests about them. After my feedback you will have to rewrite half the code, and rewrite almost all the tests. That's bad. Really bad. We are a really small team and can't waste our development time like that. Slows everything down, and also blocks the frontend who is waiting for the backend's implementation to proceed.

-  I spent ~90 mins doing a very careful review to make sure I spot everything and explain what needs to be changed and what is okay. We all know I am a geek and all I do is sit on the pc and code all day, but even my time is limited and precious. So when I have to spend so much time in a review it's taking a toll on my productivity too. And that also decreases our productivity as a team.
  
I want to come to a point where I can trust you a lot more and can spend less time in reviews. Approving with just a quick look is the goal. Please help me achieve this goal.

File: rotkehlchen/externalapis/cryptocompare.py
Comment (vnavascues) [reply rotkehlchen/externalapis/cryptocompare.py:483]: Because `query_endpoint_price` returns `Dict[str,Any]`, I want to return `Price` and `self._special_case_handling` for `query_endpoint_price` returns a dict. I didn't want to mess up with the latter.
If I'm allowed to return a Price here (like the `query_endpoint_pricehistorical` case) I will change the function name and remove the wrapper.

Comment (LefterisJP) [reply rotkehlchen/externalapis/cryptocompare.py:483]: Sure you are. Just check where what you are working on is used. This is only used here. There is no need to keep the Dict[str, Any] return. Why would you need to do that?

File: rotkehlchen/tests/unit/test_price_historian.py
Comment (vnavascues) [reply rotkehlchen/tests/unit/test_price_historian.py]: Yes, only for this test. I've renamed it as `fake_price_historian`.

File: rotkehlchen/tests/unit/test_price_historian.py
Comment (vnavascues) [reply rotkehlchen/tests/unit/test_price_historian.py]: I was mocking the `has_fiat_to_fiat` properties not the inquirer function. I've just removed the whole oracle properties abstraction and deleted the tests.

**Review by vnavascues — COMMENTED** [review]
I've applied the requested changes and simplified quite a lot the implementation by setting `oracle` and `oracle_instances` after the creation of both singletons (in the same way ethereum is injected in Inquirer).

Let's see how the APi tests behave. Unit passed.

File: rotkehlchen/tests/unit/test_inquirer.py
Comment (LefterisJP) [inline rotkehlchen/tests/unit/test_inquirer.py:172]: This is not testing the actual code but just the fixture. This is  a useless test.

File: rotkehlchen/tests/unit/test_inquirer.py
Comment (LefterisJP) [inline rotkehlchen/tests/unit/test_inquirer.py:184]: This is not testing the actual code but just the fixture's parametrization. This is a useless test. If you want to test that `set_order` works correctly then make a test where that function is explicitly called on a fresh inquirer.

File: rotkehlchen/tests/unit/test_price_historian.py
Comment (LefterisJP) [inline rotkehlchen/tests/unit/test_price_historian.py:56]: same comment as before for a useless test. You are testing your fixture here.

File: rotkehlchen/tests/unit/test_price_historian.py
Comment (LefterisJP) [inline rotkehlchen/tests/unit/test_price_historian.py:70]: same comment as before for a useless test. You are testing your fixture parametrization here

File: rotkehlchen/tests/unit/test_price_historian.py
Comment (LefterisJP) [inline rotkehlchen/tests/unit/test_price_historian.py:87]: question: The price here can be anything right? No actual calls happen to any oracle unless I misunderstand something.

**Review by LefterisJP — COMMENTED** [review]
All right so much better already. Left a few more comments. Mostly minor I believe.

One thing you should definitely do is open an issue for the creation of a common interface for price oracles.

File: rotkehlchen/tests/unit/test_price_historian.py
Comment (vnavascues) [reply rotkehlchen/tests/unit/test_price_historian.py:87]: yes, it can be anything as there are no calls to any other function (otherwise the test would fail as the oracle instances are magic mocks without any return value)

Comment (LefterisJP) [reply rotkehlchen/tests/unit/test_price_historian.py:87]: all right thanks.

**Review by LefterisJP — APPROVED** [review]
LGTM assuming tests pass. I only checked the last commit but should be sufficient.

**Comment by vnavascues** [comment]
> LGTM assuming tests pass. I only checked the last commit but should be sufficient.

There has been an exception setting up `rotkehlchen_api_server_with_exchanges` for `tests/api/test_trades.py::test_query_trades`, requesting Kraken asset pairs has failed. I assume we can merge.

**Pull request merged**
